SQL: Add string to interval converters

bpintea opened the pull request:
This PR switches the parameter conversion from QL's to SQL
converter.
It then checks if the returned converter is non-null (i.e. a conversion
is possible) and generates an appropriate exception otherwise.
It also adds checks again the validity of the parameter's data type (i.e.
the declared data type is resolved to an ES or Java type.

Finally, keyword to interval converters are added, so that the interval
types can be accepted as parameters.

Closes #45915

bpintea committed:
Add string to interval converters

This commit switches the parameter conversion from QL's to SQL
converter.
It then checks if the returned converter is non-null (i.e. a conversion
is possible) and generates an appropriate exception otherwise.
It also adds checks again the validity of the paramter's data type (i.e.
the declared data type is resolved to an ES or Java type.

Finally, keyword to interval converters are added, so that the interval
types can be accepted as paramters.

elasticmachine commented:
Pinging @elastic/es-search (:Search/SQL)

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/SqlDataTypeConverter.java:
```diff
@@ -382,10 +399,13 @@ public static Object convert(Object value, DataType dataType) {
-        TIME_TO_BOOLEAN(fromTime(value -> value != 0));
+        TIME_TO_BOOLEAN(fromTime(value -> value != 0)),
+
+        STRING_ISO8601_PERIOD_TO_TEMPORAL(fromString(Period::parse, "duration")),
```
Shouldn't "duration" be "interval" since this is used in the error message if the conversion is not possible?

matriv replied:
Also maybe use a static string variable.

bpintea replied:
Did both, thanks.
(I've just stayed consistent with the local style previously, that uses "hard-coded" strings.)

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java:
```diff
@@ -717,7 +720,10 @@ public Literal visitParamLiteral(ParamLiteralContext ctx) {
         try {
-            return new Literal(source, converterFor(sourceType, dataType).convert(param.value), dataType);
+            if (SqlDataTypeConverter.canConvert(sourceType, dataType) == false) {
+                throw new ParsingException(source, "Can not cast value [{}] as [{}]", param.value, dataType);
```
I think we should rephrase here, the `cast` ca be misleading since there is not actual `CAST(<xxx>  AS <yyy>)`.
Maybe a message along the lines of: `Cannot bind value [{}] of type [{}] to parameter of type [{}].

Please also add a test case for this.

bpintea replied:
Thanks. I've added a test and updated the message.
I've kept the `cast` though to stay consistent with the (pre-existing) message that's return when a converter actually fails to convert a value (vs. no converter exists, as in this case). 
Besides, the confusion should be minimal, type "casting" is a coined term, `CAST` function is a very specific functionality.

matriv replied:
Since the error message refers to binding a value to a param `?`, 
imho, `bind` sounds better than `cast`.

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/type/SqlDataTypeConverterTests.java:
```diff
@@ -590,6 +599,25 @@ public void testConversionToByte() {
+        assertEquals(converterFor(KEYWORD, INTERVAL_DAY_TO_MINUTE).convert("P1DT1H1M"), Duration.ofMinutes((24 + 1) * 60 + 1));
+        assertEquals(converterFor(KEYWORD, INTERVAL_DAY_TO_SECOND).convert("P1DT1.1S"), Duration.ofSeconds(24 * 3600 + 1, 100_000_000  ));
+
+        assertEquals(converterFor(KEYWORD, INTERVAL_MINUTE_TO_SECOND).convert("PT1M1S"), Duration.ofSeconds(60 + 1));
```
Could you please also some test(s) for failed conversions?

bpintea replied:
I've added one. Not sure if it's worth adding more, since this simply tests the Duration/Period parsers.

matriv reviewed (commented):
Looks good, left some comments.

bpintea committed:
address review comments

- pull hard-coded strings into final vars;
- improve error messages;
- add more tests.

bpintea committed:
address reviews

- add hunk that didn't make it into staged changes.

matriv reviewed (approved):
LGTM

astefan reviewed (approved):
LGTM. But, in general, I would prefer describing and reproducing an issue from  the client (the ODBC driver) point of view rather  than using the REST layer for it.

costin reviewed (changes requested):
The issue with the current PR is that, by enhancing the type conversion, casting a string to an _some_ interval types are supported.

So `CAST('someString' AS INTERVAL_YEAR)` is allowed. When it shouldn't.
A better approach would be to move this type conversion on the rest layer, closer to the source.
An even better one would be to reuse the infrastructure from JDBC and use its type names, namely `java.time.Duration` and `java.time.Period` on the ODBC side.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java:
```diff
@@ -717,7 +720,11 @@ public Literal visitParamLiteral(ParamLiteralContext ctx) {
         // otherwise we need to make sure that xcontent-serialized value is converted to the correct type
         try {
-            return new Literal(source, converterFor(sourceType, dataType).convert(param.value), dataType);
+            if (SqlDataTypeConverter.canConvert(sourceType, dataType) == false) {
```
Sure will do. 
But for the future, just to understand the reason: is it because no other `canConvert` (like `ql`'s) should be used in the file and a static import would make that more obvious? (Asking since the function is used currently just once in the entire file.)

bpintea commented:
I will look into the suggested alternatives.  But just to understand your guidelines:

> So `CAST('someString' AS INTERVAL_YEAR)` is allowed. When it shouldn't.

Since I believe that an invalid literal conversion would fail fast on folding (i.e. it wouldn't be "allowed"), did you actually mean the example to use a field (like `CAST someColumn AS INTERVAL_YEAR`) in which case the query would indeed be "allowed" to be passed to ES? If so, wouldn't this be similar to, for instance, an integer casting (like `CAST (someColumn AS INT)`?

> An even better one would be to reuse the infrastructure from JDBC and use its type names, namely java.time.Duration and java.time.Period[...].

We could make use of Java's types for intervals, true, but wouldn't that be inconsistent with what the column types are named in results? Ex. `interval_year` vs. `period`.

>  [...] on the ODBC side

Should this functionality - passing INTERVAL parameters - be client-independent? The INTERVAL types aren't ODBC specific; but indeed, we'd need to change that client too if we were to change the parameter type names.

bpintea commented:
PR will be split and modified to answer feedback around general exposure of the interval types raised in https://github.com/elastic/elasticsearch/pull/51945#pullrequestreview-355565888
